src/test/scala/org/apache/spark/deploy/armada/e2e/TestOrchestrator.scala

```diff
@@ -0,0 +1,358 @@
+    masterUrl: String,
+    lookoutUrl: String,
+    scalaVersion: String,
+    sparkVersion: String,
+    sparkConfs: Map[String, String] = Map.empty,
+    assertions: Seq[TestAssertion] = Seq.empty
+)
+
+/** Manages test isolation with unique namespace and queue per test. Ensures cleanup of resources
+  * after test completion.
+  */
+case class TestContext(
+    testName: String,
+    testId: String = UUID.randomUUID().toString.take(8)
+) {
+  val namespace: String   = s"e2e-test-$testId"
+  val queueSuffix: String = testId
+
+  def labels: Map[String, String] = Map(
+    "test-id"        -> testId,
+    "test-name"      -> testName,
+    "test-framework" -> "armada-spark-e2e"
+  )
+}
+
+case class TestResult(
+    jobSetId: String,
+    queueName: String,
+    status: JobSetStatus,
+    assertionResults: Map[String, AssertionResult] = Map.empty
+)
+
+class AssertionFailedException(message: String) extends Exception(message)
+
+/** Orchestrates end-to-end testing of Spark jobs on Armada.
+  *
+  * This class manages the complete lifecycle of E2E tests including:
+  *   - Queue creation and management
+  *   - Docker-based Spark job submission
+  *   - Job monitoring and status tracking
+  *   - Runtime assertion execution (e.g., verifying ingress creation)
+  *   - Comprehensive logging of all test stages
+  *
+  * Tests can include assertions that run while the job is executing, allowing verification of
+  * runtime resources like Kubernetes ingresses. Failed assertions will cause the test to fail.
+  *
+  * @param armadaClient
+  *   Client for Armada operations (queue management, job watching)
+  * @param k8sClient
+  *   Client for Kubernetes operations (pod/ingress verification)
+  */
+class TestOrchestrator(
+    armadaClient: ArmadaOperations,
+    k8sClient: K8sOperations
+)(implicit ec: ExecutionContext) {
+
+  private val jobSubmitTimeout = 30.seconds
+  private val jobWatchTimeout  = 240.seconds
+
+  private def delay(duration: Duration): Future[Unit] = {
+    val promise = Promise[Unit]()
+    val timer   = new java.util.Timer()
+    timer.schedule(
+      new java.util.TimerTask {
+        def run(): Unit = {
+          promise.success(())
+          timer.cancel()
+        }
+      },
+      duration.toMillis
+    )
+    promise.future
+  }
+
+  def runTest(name: String, config: TestConfig): Future[TestResult] = {
+    val context = TestContext(name)
+    val testJobSetId =
+      s"e2e-${name.toLowerCase.replaceAll("[^a-z0-9]", "-")}-${System.currentTimeMillis()}"
+    val queueName = s"${config.baseQueueName}-${context.queueSuffix}"
+
+    println(s"* Starting E2E Test: $name")
+    println(s"** Test ID: ${context.testId}")
+    println(s"** Namespace: ${context.namespace}")
+    println(s"** Queue: $queueName")
+
+    val resultFuture = for {
+      _ <- k8sClient.createNamespace(context.namespace)
+      _ <- armadaClient.ensureQueue(queueName)
+      _ <- submitJob(testJobSetId, queueName, name, config, context)
```

Comment:
since these 3 statements are not creating intermediate values, do they need to be in the for comprehension?